### PR TITLE
Transit: backup/restore

### LINKED ISSUE #2667: unable to import exported transit key
[Transit Read Key](https://www.vaultproject.io/api/secret/transit/index.html#read-key-1) allows Transit endpoint to output the encryption/hmac/signing key used, but there seems to be no equivalent endpoint where these keys can be imported/restored into a Vault instance.

From [Transit Docs](https://www.vaultproject.io/docs/secrets/transit/index.html):
> The primary use case for transit is to encrypt data from applications while still storing that encrypted data in some primary data store.

So if my data is transit-encrypted and stored permanently in a data store, and I lose my Vault instance, and have exported my keys using the above method, then I will have to find a third-party library to decrypt my data using the keys... ?

### PR DESCRIPTION
**Pull request by vishalnayak**
Resolves #2667

### CODE REVIEW

File: builtin/logical/transit/path_backup.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	keyBackup, err := p.Backup(req.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"backup": keyBackup,
```
Comment (jefferai) [inline builtin/logical/transit/path_backup.go]: Base64-encode this so the input on restore matches the output of backup.

File: builtin/logical/transit/path_restore.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+	"github.com/hashicorp/vault/helper/jsonutil"
+	"github.com/hashicorp/vault/helper/keysutil"
+	"github.com/hashicorp/vault/logical"
+	"github.com/hashicorp/vault/logical/framework"
+)
+
+func (b *backend) pathRestore() *framework.Path {
+	return &framework.Path{
+		Pattern: "restore/?$",
```
Comment (jefferai) [inline builtin/logical/transit/path_restore.go]: I would make an optional name to allow them to restore to a different name if desired.

File: builtin/logical/transit/path_restore.go [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+	var keyData keysutil.KeyData
+	keyData.Policy = &keysutil.Policy{
+		Keys: keysutil.KeyEntryMap{},
+	}
+	err = jsonutil.DecodeJSON(backupBytes, &keyData)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (jefferai) [inline builtin/logical/transit/path_restore.go]: Here, update policy name if defined.

File: builtin/logical/transit/path_backup.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	if lock != nil {
+		defer lock.Unlock()
+	}
+	if err != nil {
+		return nil, err
+	}
+	if p == nil {
+		return logical.ErrorResponse(fmt.Sprintf("invalid key %q", name)), nil
+	}
+
```
Comment (jefferai) [inline builtin/logical/transit/path_backup.go]: Add a flag to trip when it's been backed up -- potentially a backup struct that has the time of last backup and the current key version of the last backup. After p.Backup runs successfully, set this information in the policy and persist and adjust keyBackup to contain it.

File: builtin/logical/transit/path_backup.go (hunk 1 above)
Comment (vishalnayak) [reply builtin/logical/transit/path_backup.go]: Fixed.

File: builtin/logical/transit/path_restore.go (hunk 2 above)
Comment (vishalnayak) [reply builtin/logical/transit/path_restore.go]: Fixed.

File: builtin/logical/transit/path_restore.go (hunk 3 above)
Comment (vishalnayak) [reply builtin/logical/transit/path_restore.go]: Done.

File: builtin/logical/transit/path_backup.go (hunk 4 above)
Comment (vishalnayak) [reply builtin/logical/transit/path_backup.go]: Post this comment, we landed on *not* having a flag and just to have the BackupInfo struct. I am persisting the backup info before the `Backup()` func returns successfully.

Done.

**Pull request merged**